src/192/main/kotlin/org/rust/ide/hints/RsInlayTypeHintsProvider.kt

```diff
@@ -0,0 +1,181 @@
+import javax.swing.JPanel
+
+class RsInlayTypeHintsProvider : InlayHintsProvider<RsInlayTypeHintsProvider.Settings> {
+    override val key: SettingsKey<Settings> get() = ourKey
+
+    override val name: String get() = "Type hints"
+
+    override val previewText: String?
+        get() = """
+            struct Foo<T1, T2, T3> { x: T1, y: T2, z: T3 }
+            
+            fn main() {
+                let foo = Foo { x: 1, y: "abc", z: true };
+            }
+            """.trimIndent()
+
+    override fun createConfigurable(settings: Settings): ImmediateConfigurable = object : ImmediateConfigurable {
+        val showForVariables = "Show for variables"
+        val showForLambdas = "Show for closures"
+        val showForIterators = "Show for iterators"
+        val showObviousTypes = "Show obvious types"
+
+        private val varField = CheckBox(showForVariables)
+        private val lambdaField = CheckBox(showForLambdas)
+        private val iteratorField = CheckBox(showForIterators)
+        private val obviousTypesField = CheckBox(showObviousTypes)
+
+        override fun createComponent(listener: ChangeListener): JPanel {
+            varField.isSelected = settings.showForVariables
+            varField.addItemListener { handleChange(listener) }
+            lambdaField.isSelected = settings.showForLambdas
+            lambdaField.addItemListener { handleChange(listener) }
+            iteratorField.isSelected = settings.showForIterators
+            iteratorField.addItemListener { handleChange(listener) }
+            obviousTypesField.isSelected = settings.showObviousTypes
+            obviousTypesField.addItemListener { handleChange(listener) }
+
+            val panel = panel {
+                row { varField(pushX) }
+                row { lambdaField(pushX) }
+                row { iteratorField(pushX) }
+                row { obviousTypesField(pushX) }
+            }
+            panel.border = JBUI.Borders.empty(5)
+            return panel
+        }
+
+        private fun handleChange(listener: ChangeListener) {
+            settings.showForVariables = varField.isSelected
+            settings.showForLambdas = lambdaField.isSelected
+            settings.showForIterators = iteratorField.isSelected
+            settings.showObviousTypes = obviousTypesField.isSelected
+            listener.settingsChanged()
+        }
+    }
+
+    override fun createSettings(): Settings = Settings()
+
+    override fun getCollectorFor(file: PsiFile, editor: Editor, settings: Settings, sink: InlayHintsSink): InlayHintsCollector? =
+        object : FactoryInlayHintsCollector(editor) {
+
+            val typeHintsFactory = RsTypeHintsPresentationFactory(factory, settings.showObviousTypes)
+
+            override fun collect(element: PsiElement, editor: Editor, sink: InlayHintsSink): Boolean {
+                if (file.project.service<DumbService>().isDumb) return false
+                if (element !is RsElement) return false
```

Comment:
I'm not sure there is a big difference because AFAIK this method only is invoked when `isDumb` is false (there are checks at call sites) and when the file's language is `RsLanguage`. I've just copied the sequence from another implementation of this method.